core/src/main/scala/nl/vroste/zio/kinesis/client/Producer.scala

```diff
@@ -104,16 +132,86 @@ final case class ProducerSettings(
   failedDelay: Duration = 100.millis,
   metricsInterval: Duration = 30.seconds,
   updateShardInterval: Duration = 30.seconds,
-  aggregate: Boolean = false,
-  allowedErrorRate: Double = 0.05,
-  shardPredictionParallelism: Int = 8
+  shardPrediction: Producer.ShardPrediction = Producer.ShardPrediction.Enabled(),
+  aggregation: Producer.Aggregation = Producer.Aggregation.Disabled,
+  batchDuration: Option[Duration] = None
 ) {
-  require(allowedErrorRate > 0 && allowedErrorRate <= 1.0, "allowedErrorRate must be between 0 and 1 (inclusive)")
+  aggregation match {
+    case Producer.Aggregation.ByPredictedShard(_) =>
+      require(shardPrediction.isEnabled, "Aggregation requires shard prediction to be enabled")
+    case Producer.Aggregation.Disabled            => ()
+  }
 }
 
 object Producer {
+
+  sealed trait Throttling extends Product with Serializable
+
+  object Throttling {
+    case object Disabled extends Throttling
+
+    /**
+     * @param allowedErrorRate
+     *   The maximum allowed rate of errors before throttling is applied
+     */
+    final case class Enabled(allowedErrorRate: Double = 0.05) extends Throttling {
+      require(allowedErrorRate > 0 && allowedErrorRate <= 1.0, "allowedErrorRate must be between 0 and 1 (inclusive)")
+    }
+
+  }
+
+  sealed trait ShardPrediction extends Product with Serializable {
+    def isEnabled: Boolean = this match {
+      case ShardPrediction.Disabled   => false
+      case _: ShardPrediction.Enabled => true
+    }
+  }
+
+  object ShardPrediction {
+    case object Disabled extends ShardPrediction
+
+    /**
+     * @param parallelism
+     *   Max number of parallel shard predictions (MD5 hashing)
+     * @param throttling
+     *   Throttling settings for per-shard throttling. If None, throttling is disabled.
+     */
+    final case class Enabled(parallelism: Int = 8, throttling: Throttling = Throttling.Enabled())
```

Comment:
Currently throttling only makes sense when shardprediction is enabled, so I moved it here. Do we have other version planned / should I make it a top-level setting instead?

Reply:
See other comment